nemo/collections/common/data/lhotse/nemo_adapters.py

```diff
@@ -274,13 +277,21 @@ def __iter__(self) -> Generator[Cut, None, None]:
         for sid in shard_ids:
             shard_manifest = self.shard_id_to_manifest[sid]
             tar_path = self.shard_id_to_tar_path[sid]
-            with tarfile.open(fileobj=open_best(tar_path, mode="rb"), mode="r|*") as tar:
-                for data, tar_info in zip(shard_manifest, tar):
-                    manifest_path = self.paths[sid] if len(self.paths) > 1 else self.paths[0]
-                    assert data["audio_filepath"] == tar_info.name, (
-                        f"Mismatched entry between JSON manifest ('{manifest_path}') and tar file ('{tar_path}'). "
-                        f"Conflicting audio file names are JSON='{data['audio_filepath']}' and TAR='{tar_info.name}'"
-                    )
+            if self.random_access:
```

Comment:
First, let's wait for PR #9187 to be merged; then after resolving conflicts, I suggest we factor the sequential/random iteration into separate generator functions like this:

```python
    def _iter_random_read(self, shard_ids: Sequence[int]) -> Generator[tuple[dict, bytes], None, None]:
        for sid in shard_ids:
            shard_manifest = self.shard_id_to_manifest[sid]
            tar_path = self.shard_id_to_tar_path[sid]
            with tarfile.open(fileobj=open_best(tar_path, mode="rb"), mode="r") as tar:
                for data in shard_manifest:
                    try:
                        tar_info = tar.getmember(data["audio_filepath"])
                        raw_audio = tar.extractfile(tar_info).read()
                        yield data, raw_audio
                    except KeyError as e:
                        manifest_path = self.paths[sid] if len(self.paths) > 1 else self.paths[0]
                        raise RuntimeError(
                            f"Mismatched entry between JSON manifest ('{manifest_path}') and tar file ('{tar_path}'). "
                            f"The following audio_filepath='{data['audio_filepath']}' was not found in the tar file."
                        ) from e

    def _iter_sequential(self, shard_ids: Sequence[int]) -> Generator[tuple[dict, bytes], None, None]:
        for sid in shard_ids:
            shard_manifest = self.shard_id_to_manifest[sid]
            tar_path = self.shard_id_to_tar_path[sid]
            with tarfile.open(fileobj=open_best(tar_path, mode="rb"), mode="r|*") as tar:
                for data, tar_info in zip(shard_manifest, tar):
                    manifest_path = self.paths[sid] if len(self.paths) > 1 else self.paths[0]
                    assert data["audio_filepath"] == tar_info.name, (
                        f"Mismatched entry between JSON manifest ('{manifest_path}') and tar file ('{tar_path}'). "
                        f"Conflicting audio file names are JSON='{data['audio_filepath']}' and TAR='{tar_info.name}'"
                    )
                    raw_audio = tar.extractfile(tar_info).read()
                    yield data, raw_audio
```

(we'll need to update this bit of code to have the changes from #9187, as I had written it before that PR was submitted)

Reply:
Done